distsql: use custom not supported error

RaduBerinde opened the pull request:
Avoid the use of `errors.Errorf` in the hot path for queries not supported by
DistSQL.

Fixes #16549.

RaduBerinde committed:
distsql: use custom not supported error

Avoid the use of `errors.Errorf` in the hot path for queries not supported by
DistSQL.

Fixes #16549.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16553)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553#-KmiVDdUAX2W8x-FRYNL:-KmiVDdUAX2W8x-FRYNM:bhfojx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebe6d011ef868e36aa8c90f92f92b8b809d12a77/pkg/sql/distsql_physical_planner.go#L317)):*
> ```Go
> 
> 	default:
> 		return 0, queryNotSupportedError(fmt.Sprintf("unsupported node %T", node))
> ```

This is the common case for `insertNode`. Somewhat unfortunate that it involves several allocations. Do we really need the detail about the node type that is unsupported? I suppose we could special case a few nodes like `insertNode`, `deleteNode` and `updateNode`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553#-:-KmiWHeA-LA4RoVhKvJ-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553#-KmiVDdUAX2W8x-FRYNL:-KmifBmAkZSyYCzV_4eC:bdlm43l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebe6d011ef868e36aa8c90f92f92b8b809d12a77/pkg/sql/distsql_physical_planner.go#L317)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is the common case for `insertNode`. Somewhat unfortunate that it involves several allocations. Do we really need the detail about the node type that is unsupported? I suppose we could special case a few nodes like `insertNode`, `deleteNode` and `updateNode`. 
</blockquote></details>

Yeah I thought about that too.. Good idea, added special case. I switched to using `fmt.Errorf` and used a cached error here to avoid an allocation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/distsql_physical_planner.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553#-KmiVDdUAX2W8x-FRYNL:-KmijdxarvUqNQGK0IV3:b-d3x6ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebe6d011ef868e36aa8c90f92f92b8b809d12a77/pkg/sql/distsql_physical_planner.go#L317)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yeah I thought about that too.. Good idea, added special case. I switched to using `fmt.Errorf` and used a cached error here to avoid an allocation.
</blockquote></details>

Actually, sorry, I kept `queryNotSupportedError` but made it a struct and used pointers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553#-KmiVDdUAX2W8x-FRYNL:-KmlQDoGByvvA1rCRGUB:bbk96au) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebe6d011ef868e36aa8c90f92f92b8b809d12a77/pkg/sql/distsql_physical_planner.go#L317)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Actually, sorry, I kept `queryNotSupportedError` but made it a struct and used pointers.
</blockquote></details>

Ok. I'm curious about the change from using a `queryNotsupportedError` value to `newQueryNotSupportedError` which returns a pointer. I believe both approaches cause an equivalent number of allocations. With the pointer you are explicitly allocating. With the value, there is an implicit allocation when it gets stored in the interface. Wondering if there was another reason for the change that I'm not seeing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/distsql_physical_planner.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553#-KmiVDdUAX2W8x-FRYNL:-KmlSdbjrC6Usi7CEJbG:b-7grzw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebe6d011ef868e36aa8c90f92f92b8b809d12a77/pkg/sql/distsql_physical_planner.go#L317)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok. I'm curious about the change from using a `queryNotsupportedError` value to `newQueryNotSupportedError` which returns a pointer. I believe both approaches cause an equivalent number of allocations. With the pointer you are explicitly allocating. With the value, there is an implicit allocation when it gets stored in the interface. Wondering if there was another reason for the change that I'm not seeing.
</blockquote></details>

Because of `mutationsNotSupportedError`. If I just did `var mutationsNotSupportedError = queryNotSupportedError`, there will still be an allocation when we return it as an `error`. I could have done `var mutationsNotSupportedError error = queryNotSupportedError` but I thought the distinction was too subtle. I noticed `fmt.Errorf` uses a struct and that seemed more sraightforward.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16553)*
<!-- Sent from Reviewable.io -->
